### ISSUE TITLE
Pyright fails to check the result returned by a function

### ISSUE THREAD

**Issue by ChenZhongPu** [microsoft/pyright#6206]
I am using Pyright in Nvim (Note that such bug cannot be reproduced in VSCode with Pylance):

<img width="1353" alt="image" src="https://github.com/microsoft/pyright/assets/2310958/3b901844-5c9a-47dc-99cc-65e7962ec058">

As we can see, Pyright does not know the `_tail` is not `None` in the `else` case. On the other hand, if we replace `if self.empty()` with `if  self._tail is None`, then Pyright will not complain "next is not a member of None".

The full code:

```python
class Node:
    def __init__(self, item, next=None):
        self.item = item
        self.next = next


class CircularLinkedList:
    def __init__(self):
        self._tail = None
        self._size = 0

    def is_empty(self):
        return self._tail is None

    def add_first(self, item):
        if self.is_empty():
            self._tail = Node(item)
            self._tail.next = self._tail
        else:
            self._tail.next = Node(item, self._tail.next)
        self._size += 1
```

**Comment by erictraut** [comment]
Please provide a self-contained minimal example in text (code) form as requested in the bug report template. Debugging issues from partial screen shots isn't really feasible.

**Comment by erictraut** [comment]
Pyright is working correctly here. This isn't a bug.

The type of `self._tail` is `Node | None`. You are attempting to access attribute `next` which would cause a runtime exception if `self._tail` is None`. Although you have performed a runtime check by calling `is_empty`, static type checkers do not perform global type analysis, nor can they generally assume that a condition that's true will remain true across function calls.

If you want this code to type check, here's the approach I'd recommend:

```python
    def add_first(self, item):
        if (tail := self._tail) is None:
            self._tail = Node(item)
            self._tail.next = None
        else:
            tail.next = Node(item, tail.next)
        self._size += 1
```

If you want to stick with your current approach, you can insert an `assert` into your code to document (and verify) your assumptions.

```python
    def add_first(self, item):
        if self.is_empty():
            self._tail = Node(item)
            self._tail.next = self._tail
        else:
            assert self._tail is not None
            self._tail.next = Node(item, self._tail.next)
        self._size += 1
```
